src/Evaluator.hs

```diff
@@ -0,0 +1,214 @@
+evalQualifiedVar (QualifiedVar modName identifier _ _) env = 
+    return $ lookupQualifiedVar modName identifier env
+evalQualifiedVar expr _ = exprTypeError expr "QualifiedVar"
+
+evalQualifiedConstructor :: Show a => Expr a -> Environment -> IO Value
+evalQualifiedConstructor (QualifiedConstructor modName constructorName _ _) env= 
+    return $ lookupQualifiedConstructor modName constructorName env
+evalQualifiedConstructor expr _ = exprTypeError expr "QualifiedConstructor"
+
+evalAnd :: Show a => Expr a -> Environment -> IO Value
+evalAnd (And leftExpr rightExpr _ _) env = do
+    leftVal <- eval leftExpr env
+    case leftVal of
+        PrimitiveVal (BoolVal False) -> return leftVal
+        PrimitiveVal (BoolVal True ) -> do
+            rightVal <- eval rightExpr env
+            case rightVal of
+                PrimitiveVal (BoolVal _) -> return rightVal
+                _                        -> evalError rightExpr "bool"
+        _ -> evalError leftExpr "Bool"
+evalAnd expr _ = exprTypeError expr "And"
+
+evalOr :: Show a => Expr a -> Environment -> IO Value
+evalOr (Or leftExpr rightExpr _ _) env = do
+    leftVal <- eval leftExpr env
+    case leftVal of
+        PrimitiveVal (BoolVal True ) -> return leftVal
+        PrimitiveVal (BoolVal False) -> do
+            rightVal <- eval rightExpr env
+            case rightVal of
+                PrimitiveVal (BoolVal _) -> return rightVal
+                _                        -> error $ show rightExpr
+        _ -> evalError leftExpr "Bool"
+evalOr expr _ = exprTypeError expr "Or"
+
+evalIf :: Show a => Expr a -> Environment -> IO Value
+evalIf (If predicate consequent alternative _ _) env = do
+    predicateVal <- eval predicate env
+    case predicateVal of
+        PrimitiveVal (BoolVal True ) -> eval consequent env
+        PrimitiveVal (BoolVal False) -> eval alternative env
+        _                            -> evalError predicate "Bool"
+evalIf expr _ = exprTypeError expr "If"
+
+evalBlock :: Show a => Expr a -> Environment -> IO Value
+evalBlock (Block exprs _ _) env = evalSequence exprs  where
+    evalSequence :: Show a => [Expr a] -> IO Value
+    evalSequence []                      = return $ PrimitiveVal $ UnitVal
+    evalSequence [expr                 ] = eval expr env
+    evalSequence (firstExpr : restExprs) = do
+        eval firstExpr env
+        evalSequence restExprs
+evalBlock expr _ = exprTypeError expr "Block"
+
+exprUnit :: Expr a -> Expr ()
+exprUnit = fmap (const ())
+
+evalLambda :: Show a => Expr a -> Environment -> IO Value
+evalLambda lambdaExpr@Lambda{} env = return
+    $ makeProcedure (exprUnit lambdaExpr) env  where
+        makeProcedure :: Expr () -> Environment -> Value
+        makeProcedure (Lambda pattern expr name _ _) env =
+            Procedure pattern expr env
+evalLambda expr _ = exprTypeError expr "Lambda"
+
+evalExprList :: Show a => [Expr a] -> Environment -> IO [Value]
+evalExprList []            env = return []
```

Comment:
Isn't it just `mapM`?

Reply:
![image](https://user-images.githubusercontent.com/33699940/88203095-129ac300-cc4a-11ea-8387-7fec1e3f3c99.png)

Reply:
![image](https://user-images.githubusercontent.com/33699940/88203353-3a8a2680-cc4a-11ea-9f4e-5fe273104188.png)